### ISSUE TITLE
How to obtain differentiable keypoints in YOLOv8 pose estimator?

### ISSUE THREAD

**Issue by mashrurmorshed** [ultralytics/ultralytics#13226]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I want to obtain differentiable keypoints to compute a custom loss on an image generated by another model.

As an example, I will just try to compute gradients with respect to an image pulled off the web:

```python
import torch
from ultralytics import YOLO
from torchvision import transforms as T
from PIL import Image
import requests
from io import BytesIO

# load yolo
device = "cuda:0"
model = YOLO("yolov8n-pose.pt").to(device)

# image_tensor is a (1, C, H, W) tensor in range [0., 1.]
image_url = "https://raw.githubusercontent.com/open-mmlab/mmpose/main/tests/data/coco/000000000785.jpg"
img = Image.open(BytesIO(requests.get(image_url).content)).convert("RGB").resize((640, 640))
img_tensor = T.ToTensor()(img).unsqueeze(0).to(device).requires_grad_(True)

result = model(img_tensor)
kpt = result[0].keypoints.xy

print(kpt.requires_grad) # gives False

# (cannot do following steps)
# kpt_loss = some_loss_function(kpt)
# gradient = torch.autograd.grad(img_tensor, kpt_loss)[0]
```

There are probably several reasons why the results don't have gradients, maybe because of converting between tensor to numpy or forced torch.inference_mode() inside the model.

To really get differentiable keypoints, I most likely need to get the raw outputs. I can do that by:

```python
out = model.model(img_tensor)
```

This produces tensors with gradients, but I am assuming I will need to do some post-processing myself to get the keypoints (while keeping gradients). `out` contains 5 tensors, in a nested structure like this:

```python
(
    torch.Size([1, 56, 8400]),
    (
        [
            torch.Size([1, 65, 80, 80]),
            torch.Size([1, 65, 40, 40]),
            torch.Size([1, 65, 20, 20])
        ],
        torch.Size([1, 51, 8400])
    )
)
```
I would like some help on how to obtain keypoints from the raw prediction.

**Comment by glenn-jocher** [comment]
Hello!

To obtain differentiable keypoints from YOLOv8's pose estimation model, you're on the right track by accessing the raw outputs directly from the model. This approach allows you to maintain the gradients necessary for custom loss computation.

Here's a brief guide on how to proceed:

1. **Extract Raw Outputs**: As you've done, use `out = model.model(img_tensor)` to get the raw model outputs.

2. **Post-Processing**: You'll need to implement the post-processing steps that convert these raw outputs into keypoints. This typically involves applying a sigmoid function to the output tensors to get the keypoint locations and applying non-maximum suppression (NMS) to refine these keypoints.

3. **Differentiable Keypoints**: Ensure that all operations in your post-processing are differentiable. PyTorch operations are generally differentiable, but be cautious with operations that might detach gradients or are not differentiable by nature.

4. **Custom Loss Function**: Once you have your differentiable keypoints, you can define your custom loss function and compute gradients as needed.

For the specific details on how to decode the raw outputs into keypoints, you might need to refer to the specific architecture details or the original implementation of the model. This often involves understanding the output tensor dimensions and how they map to keypoints.

If you need further assistance or have more questions, feel free to ask!

**Comment by mashrurmorshed** [comment]
Hi Glenn, are your answers ChatGPT generated by any chance?

I've figured out what to do.

```python
from ultralytics.utils import ops

preds = model.model(img_tensor)

preds_nms = ops.non_max_suppression(
    preds,
    model.predictor.args.conf,
    model.predictor.args.iou,
    agnostic=model.predictor.args.agnostic_nms,
    max_det=model.predictor.args.max_det,
    classes=model.predictor.args.classes,
    nc=len(model.predictor.model.names),
)

pred_kpts = preds_nms[0][:, 6:].view(1, *model.model.kpt_shape)
```

This gives the correct (17, 3) tensor for keypoints, and importantly, they are still on the computation graph.

**Comment by glenn-jocher** [comment]
Hello!

That's fantastic to hear you've figured out the solution! Your approach to extracting keypoints while keeping them on the computation graph looks solid. Thank you for sharing your code snippet; it's a great help for others who might be facing similar challenges.

If you have any more questions or need further assistance as you continue with your project, feel free to reach out. Happy coding! 😊
